From 4f9169c70c83291e719f20269ee429093492798d Mon Sep 17 00:00:00 2001
From: Georg Sauthoff <mail@georg.so>
Date: Sun, 29 Jul 2018 16:58:09 +0200
Subject: [PATCH] Guard some imports and tests for Fedora 27

such that `python3 setup.py test` executes successfully.

Also switch to secp224r1 as unknown curve. In contrast to secp112r1 it's
more widely available (e.g. also Fedora 27/28). CentOS 7 doesn't have it
or any other curve that is unknown to asyncssh such that the test is
skipped in such environments.
---
 tests/sspi_stub.py       |  8 ++++++--
 tests/test_public_key.py | 15 ++++++++++-----
 2 files changed, 16 insertions(+), 7 deletions(-)

diff --git a/tests/sspi_stub.py b/tests/sspi_stub.py
index b2748c2..c176c30 100644
--- a/tests/sspi_stub.py
+++ b/tests/sspi_stub.py
@@ -12,8 +12,12 @@
 
 """Stub SSPI module for unit tests"""
 
-from asyncssh.gss_win32 import ASC_RET_INTEGRITY, ISC_RET_INTEGRITY
-from asyncssh.gss_win32 import SECPKG_ATTR_NATIVE_NAMES, SSPIError
+import sys
+# on non-windows systems, work around setup.py test discovery
+# importing everything
+if sys.platform == 'win32':
+    from asyncssh.gss_win32 import ASC_RET_INTEGRITY, ISC_RET_INTEGRITY
+    from asyncssh.gss_win32 import SECPKG_ATTR_NATIVE_NAMES, SSPIError
 
 from .gss_stub import step
 
diff --git a/tests/test_public_key.py b/tests/test_public_key.py
index 3c988e3..843af93 100644
--- a/tests/test_public_key.py
+++ b/tests/test_public_key.py
@@ -26,6 +26,7 @@ from pathlib import Path
 import shutil
 import subprocess
 import sys
+import unittest
 
 import asyncssh
 
@@ -1968,11 +1969,15 @@ class _TestPublicKeyTopLevel(TempDirTestCase):
                         '-param_enc explicit' % curve)
                     asyncssh.read_private_key('priv')
 
-            with self.subTest('Import EC key with unknown explicit parameters'):
-                run('openssl ecparam -out priv -noout -genkey -name secp112r1 '
-                    '-param_enc explicit')
-                with self.assertRaises(asyncssh.KeyImportError):
-                    asyncssh.read_private_key('priv')
+    @unittest.skipIf(b'secp224r1' not in run('openssl ecparam -list_curves'),
+            "this openssl doesn't support secp224r1")
+    @unittest.skipIf(not _openssl_available, "openssl isn't available")
+    def test_ec_explicit_unk(self):
+        """Import EC key with unknown explicit parameters"""
+        run('openssl ecparam -out priv -noout -genkey -name secp224r1 '
+            '-param_enc explicit')
+        with self.assertRaises(asyncssh.KeyImportError):
+                asyncssh.read_private_key('priv')
 
     def test_generate_errors(self):
         """Test errors in private key and certificate generation"""
-- 
2.14.4

